challenges/systems/challenge-2.md

```diff
@@ -0,0 +1,298 @@
+thinking, for example:
+
+```
+  // TODO: Add configuration system.
+  // Consider using CLI library to support both
+  // environment variables and reasonable default values,
+  // for example https://github.com/alecthomas/kingpin
+```
+
+Comments like this one are really helpful to us. They save yourself a lot of
+time and demonstrate that you've spent time thinking about this problem and
+provide a clear path to a solution.
+
+Consider making other reasonable trade-offs. Make sure you communicate them to
+the interview team.
+
+Here are some other trade-offs that will help you to spend less time on the
+task:
+
+* Do not implement a system that scales or is highly performing. Describe which
+  performance improvements you would add in the future.
+* It is OK if the system is not highly available. Write down how you would make
+  the system highly available and why your system is not.
+
+## Pitfalls and Gotchas
+
+To help you out, we've composed a list of things that previously resulted in a
+no-pass from the interview team:
+
+* Scope creep. Candidates have tried to implement too much and ran out of time
+  and energy. To avoid this pitfall, use the simplest solution that will work.
+  Avoid writing too much code. We've seen candidates' code introducing caching
+  and making many mistakes in the caching layer validation logic. Not having
+  caching would have solved this problem.
+* Data races. We will scan the code with a race detector and do our best to
+  find data races in the code. Avoid global state as much as possible; if using
+  global state, write down a good description why it is necessary and protect it
+  against data races.
+* Deadlocks. When using mutexes, channels or any other synchronization
+  primitives, make sure the system won't deadlock. We've seen candidates' code
+  holding a mutex and making a network call without timeouts in place. Be extra
+  careful with networking and sync primitives.
+* Unstructured code. We've seen candidates leaving commented chunks of code,
+  having one large file with all the code, not having code structure at all.
+* Not communicating. Some candidates have submitted all their code to the
+  master branch without raising pull requests, which does not give us the
+  ability to provide feedback on the various implementation phases. We are a
+  distributed team, so structured, asynchronous communication is critical to us.
+* Implementing custom security algorithms/authentication schemes is always a
+  bad idea unless you are a trained security researcher/engineer. It is
+  definitely a bad idea for this task - try to stick to industry proven security
+  methods as much as possible.
+
+## Questions
+
+It is OK to ask the interview team questions. Some folks stay away from asking
+questions to avoid appearing less experienced, so we provide examples of
+questions to ask and questions we expect candidates to figure out on their own.
```

Comment:
```suggestion
It is okay to ask the interview team questions. Some folks stay away from asking
questions to avoid appearing less experienced, so we provide examples of
questions to ask and questions we expect candidates to figure out on their own.
```